### PR TITLE
[risk=low][RW-11491] Automate deployments to the Stable environment

### PR DESCRIPTION
**Pull request by Peter-Lavigne**
Before this PR, developers deployed the `stable` environment as part of the [release process](https://docs.google.com/document/d/17YVCqu4BhVV8ai3OutHx8u0m6V5NkdM5H9kLNxTPDdo/edit?tab=t.0#heading=h.8kph05q8yxo7). After this PR, `stable` is automatically deployed after `staging` tests pass and the developer manually approves the deployment.

I added a manual approval step because there will still be manual steps between "deploy staging" and "deploy stable" such as assigning the release ticket to yourself. These tasks can be automated in the future.

This is a "low" risk PR because it provides CircleCI with credentials for the `deploy@all-of-us-rw-stable.iam.gserviceaccount.com` service account. I personally suspect this is safer, security-wise, than our current approach because I think CircleCI can keep one secret values safer than ~10 release engineers running this on their local machines. Some evidence of this is the fact that we have multiple long-running _prod_ keys for these accounts active right now, when we should have zero, which I think indicates that errors on developer machines can lead to keys not being revoked. I've asked security about this [here](https://pmi-engteam.slack.com/archives/C0491NK4N1G/p1702397525640409).

I have attemped to implement this in a way that makes automating pre-prod and prod trivial.

Validating this is tricky and must be done manually. I've been deploying to CircleCI by tagging a branch with the following CircleCI config changes:

```
 filter-stable-deployment-manual-verification-tags: &filter-stable-deployment-manual-verification-tags
   filters:
     branches:
       ignore: /.*/
     tags:
       only: /^-manual-verification-[0-9]+$/
creating tags locally

...

 deploy-stable:
   jobs:
     - deploy-to-stable:
         <<: *filter-stable-deployment-manual-verification-tags
```

You can see an example deployment [here](https://app.circleci.com/pipelines/github/all-of-us/workbench/30303/workflows/3cc76336-f0a2-4bac-b5a3-e7bdaa66874c).

---
**PR checklist**

- [x] I have included an issue ID or "no ticket" in the PR title as outlined in [CONTRIBUTING.md](https://github.com/all-of-us/workbench/blob/main/.github/CONTRIBUTING.md).
- [x] I have included a risk tag of the form `[risk=no|low|moderate|severe]` in the PR title as outlined in [CONTRIBUTING.md](https://github.com/all-of-us/workbench/blob/main/.github/CONTRIBUTING.md).
- [x] I have manually tested this change and my testing process is described above.
- [x] This change includes appropriate automated tests, and I have documented any behavior that cannot be tested with code.
- [x] I have added explanatory comments where the logic is not obvious.
- One or more of the following is true:
  - [x] This change is intended to complete a JIRA story, so I have checked that all AC are met for that story.
  - [ ] This change fixes a bug, so I have ensured the steps to reproduce are in the Jira ticket or provided above.
  - [ ] This change impacts deployment safety (e.g. removing/altering APIs which are in use), so I have documented the impacts in the description.
  - [ ] This change includes a new feature flag, so I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later.
  - [ ] This change modifies the UI, so I have taken screenshots or recordings of the new behavior and notified the PO and UX designer in [Slack](https://pmi-engteam.slack.com/archives/C02MWP2RN5P).
  - [ ] This change modifies API behavior, so I have run the relevant E2E tests locally because API changes are not covered by our PR checks.
  - [ ] None of the above apply to this change.

### CODE REVIEW

File: ci/Dockerfile.circle_build
Comment (Peter-Lavigne) [inline ci/Dockerfile.circle_build:3]: Discovered as part of debugging [this issue](https://pmi-engteam.slack.com/archives/CHRN2R51N/p1701984349958949). It turns out changing this file was not necessary; see the `deps.sh` change below.

File: ui/src/deps/deps.sh
Comment (Peter-Lavigne) [inline ui/src/deps/deps.sh:13]: I think this is a better way to solve the issue described [here](https://pmi-engteam.slack.com/archives/C0189PPPKFF/p1694032310063459). Instead of globally installing `openapi-generator-cli` in the Dockerfile (which was breaking stable deployments), run yarn install before installing other dependencies. This installs `openapi-generator-cli` locally instead of globally.

File: ci/activate_creds.sh
Comment (Peter-Lavigne) [inline ci/activate_creds.sh:9]: Allow overriding env var names, but keep the defaults as `GCLOUD_CREDENTIALS` and `GCLOUD_CREDENTIALS_KEY` so that other uses of this script don't break. It's possible we can remove the need for defaults entirely, but I'm not confident I won't break an unrelated workflow. I'd need to research this more.

File: .circleci/config.yml
Comment (Peter-Lavigne) [inline .circleci/config.yml:942]: See https://circleci.com/blog/deploying-with-approvals/

File: .circleci/config.yml
Comment (Peter-Lavigne) [inline .circleci/config.yml:174]: I created two new CircleCI secrets containing the values for the `stable` deploy service account, using the same method we use for `GCLOUD_CREDENTIALS` and `GCLOUD_CREDENTIALS_KEY` as described in the [oncall playbook](https://docs.google.com/document/d/1hbYNWzdxuRybTfW3M70hWZfr_TXRVd5rOMR9kZtwAFY/edit?tab=t.0#heading=h.676040liklzd). Once this is merged, I'll add a new section to that playbook.

**Review by Qi77Qi — APPROVED** [review]
looks great!

👍 once we get security approval

**Comment by Peter-Lavigne** [comment]
The security team would like me to write a proposal outlining the approach. Approval may take awhile so I'm going to revoke the service account key in the meantime. Manually testing this will no longer work unless you rotate the keys.

File: ci/Dockerfile.circle_build
Comment (Peter-Lavigne) [reply ci/Dockerfile.circle_build:3]: This comment thread has been moved [here](https://github.com/all-of-us/workbench/pull/8239#discussion_r1431926512).

File: ui/src/deps/deps.sh
Comment (Peter-Lavigne) [reply ui/src/deps/deps.sh:13]: This comment thread has been moved [here](https://github.com/all-of-us/workbench/pull/8239#discussion_r1431926648).

**Comment by Peter-Lavigne** [comment]
I split https://github.com/all-of-us/workbench/pull/8239 out of this PR because those commits do not require security approval.

File: ci/Dockerfile.circle_build
Comment (jmthibault79) [reply ci/Dockerfile.circle_build:3]: Stolen for #8238 :)

Comment (Peter-Lavigne) [reply ci/Dockerfile.circle_build:3]: @jmthibault79 This line was merged in https://github.com/all-of-us/workbench/pull/8239, so you'll need to rebase.

**Comment by Peter-Lavigne** [comment]
I'm closing this PR, see details [here](https://precisionmedicineinitiative.atlassian.net/browse/RW-11491?focusedCommentId=189571).

**Comment by jmthibault79** [comment]
Rebased on to latest main as #8328 but preserving this version as context for the comments

**Pull request closed without merging**
